app/assets/javascripts/oxalis/view/version_view.js

```diff
@@ -17,14 +16,25 @@ import { setAnnotationAllowUpdateAction } from "oxalis/model/actions/annotation_
 import { revertToVersion } from "oxalis/model/sagas/update_actions";
 import { pushSaveQueueAction, setVersionNumberAction } from "oxalis/model/actions/save_actions";
 import { enforceSkeletonTracing } from "oxalis/model/accessors/skeletontracing_accessor";
+import type { OxalisState, SkeletonTracingType } from "oxalis/store";
+import type { UpdateAction } from "oxalis/model/sagas/update_actions";
+
+type AddToValueFn = <T>(
+  T,
+) => {
+  name: $PropertyType<T, "name">,
+  value: { ...$PropertyType<T, "value">, actionTimestamp: number, version: number },
```

Comment:
In general, this already looks very nice. Unfortunately, flow cannot understand that `$PropertyType<T, "name">` and `$PropertyType<T, "value">` should always refer to the same Union-part. That's why it thinks that name can be "updateTracing or deleteNode or ..." while simultaneously thinking "value can be X or Y or Z". It doesn't derive the constraint "if name is updateTracing, value is X".

I didn't find a nice way to express what we want, which is why I ended up defining a `ServerUpdateAction` which is a union of `AsServerType<UpdateTracing`-style parts. Additionally, it was necessary to add extra type-checks, since the typings of `_.groupBy` are not expressive enough (not sure whether they can be improved or if flow is too limited here, too).